package/src/lsif.ts

```diff
@@ -0,0 +1,229 @@
+        }
+        return await response.json()
+    })()
+
+    lsifDocs.set(doc.uri, hasLSIFPromise)
+
+    return hasLSIFPromise
+}
+
+async function hover(
+    doc: sourcegraph.TextDocument,
+    position: sourcegraph.Position
+): Promise<sourcegraph.Hover | null> {
+    const hover: LSP.Hover | null = await queryLSIF({
+        doc,
+        method: 'hover',
+        path: pathFromDoc(doc),
+        position,
+    })
+    if (!hover) {
+        return null
+    }
+    return convertHover(sourcegraph, hover)
+}
+
+async function definition(
+    doc: sourcegraph.TextDocument,
+    position: sourcegraph.Position
+): Promise<sourcegraph.Definition | null> {
+    const body: LSP.Location | LSP.Location[] | null = await queryLSIF({
+        doc,
+        method: 'definitions',
+        path: pathFromDoc(doc),
+        position,
+    })
+    if (!body) {
+        return null
+    }
+    const locations = Array.isArray(body) ? body : [body]
+    return convertLocations(
+        sourcegraph,
+        locations.map((definition: LSP.Location) => ({
+            ...definition,
+            uri: setPath(doc, definition.uri),
+        }))
+    )
+}
+
+async function references(
+    doc: sourcegraph.TextDocument,
+    position: sourcegraph.Position
+): Promise<sourcegraph.Location[] | null> {
+    const locations: LSP.Location[] | null = await queryLSIF({
+        doc,
+        method: 'references',
+        path: pathFromDoc(doc),
+        position,
+    })
+    if (!locations) {
+        return []
+    }
+    return convertLocations(
+        sourcegraph,
+        locations.map((reference: LSP.Location) => ({
+            ...reference,
+            uri: setPath(doc, reference.uri),
+        }))
+    )
+}
+
+export type Maybe<T> = { value: T } | undefined
+
+export const wrapMaybe = <A extends any[], R>(
+    f: (...args: A) => Promise<R>
+) => async (...args: A): Promise<Maybe<R>> => {
+    const r = await f(...args)
+    return r !== undefined ? { value: r } : undefined
+}
+
+export function asyncWhen<A extends any[], R>(
+    asyncPredicate: (...args: A) => Promise<boolean>
+): (f: (...args: A) => Promise<R>) => (...args: A) => Promise<Maybe<R>> {
+    return f => async (...args) =>
+        (await asyncPredicate(...args))
+            ? { value: await f(...args) }
+            : undefined
+}
+
+export function when<A extends any[], R>(
+    predicate: (...args: A) => boolean
+): (f: (...args: A) => Promise<R>) => (...args: A) => Promise<Maybe<R>> {
+    return f => async (...args) =>
+        predicate(...args) ? { value: await f(...args) } : undefined
+}
+
+export const asyncFirst = <A extends any[], R>(
+    fs: ((...args: A) => Promise<Maybe<R>>)[],
+    defaultR: R
+) => async (...args: A): Promise<R> => {
+    for (const f of fs) {
+        const r = await f(...args)
+        if (r !== undefined) {
+            return r.value
+        }
+    }
+    return defaultR
+}
+
+export function initLSIF() {
```

Comment:
Could it be that there is no TSLint setup in this repo? Because this function doesn't have a return type annotation, which would be rejected by our TSLint config (and also would have made it a lot easier to understand this code, as a reviewer)

Reply:
WIP on tslint in https://github.com/sourcegraph/sourcegraph-basic-code-intel/pull/125